### PR TITLE
[Merged by Bors] - feat(RingTheory/LaurentSeries): add properties of the X-adic valuation on Laurent series

### PR DESCRIPTION
**Pull request by faenuccio**
Add some properties conneccting the $X$-adic valuation of a Laurent series to the vanishing of its coefficients, together with explicit values of the valuation of some basic Laurent series.

Co-authored-by: María Inés de Frutos-Fernández @mariainesdff 

- [x] depends on: #13064

### CODE REVIEW

**Comment by ScottCarnahan** [comment]
Have you considered using `AddVal` from `Mathlib.RingTheory.HahnSeries.Summable`?  It looks like the `order` API from Hahn series can make some of your proofs shorter.

**Comment by faenuccio** [comment]
> Have you considered using `AddVal` from `Mathlib.RingTheory.HahnSeries.Summable`? It looks like the `order` API from Hahn series can make some of your proofs shorter.

Oh no, I had briefly had a look and then decided that since I was going for a `MulVal` they were unrelated. I will check it and try to simply my proofs - thanks!

**Comment by faenuccio** [comment]
> Have you considered using `AddVal` from `Mathlib.RingTheory.HahnSeries.Summable`? It looks like the `order` API from Hahn series can make some of your proofs shorter.

I have had a look but it seems to me that since most of the material about `AddVal` concerns... additive valuations, there is not much I can gain: I would at any rate need to set-up a dictionary. Or am I missing something, do you have explicit instances in mind where this could help?

File: Mathlib/RingTheory/LaurentSeries.lean [hunk 1]
```diff
@@ -27,10 +27,22 @@ import Mathlib.RingTheory.Localization.FractionRing
...
 the underlying `RatFunc.coeAlgHom`.
-* Study of the X-Adic valuation on the ring of Laurent series over a field
+* Study of the `X`-Adic valuation on the ring of Laurent series over a field
 
 ## Main Results
 * Basic properties of Hasse derivatives
+### About the `X`-Adic valuation:
+* The (integral) valuation of a power series is the order of the first non-zero coefficient, see
+`intValuation_le_iff_coeff_lt_eq_zero`.
+* The valuation of a laurent series is the order of the first non-zero coefficient, see
```
Comment (faenuccio) [inline Mathlib/RingTheory/LaurentSeries.lean:37]:
```suggestion
* The valuation of a Laurent series is the order of the first non-zero coefficient, see
```

File: Mathlib/RingTheory/LaurentSeries.lean [hunk 2]
```diff
@@ -483,4 +490,168 @@ theorem intValuation_X : (idealX K).intValuationDef X = ↑(Multiplicative.ofAdd
...
+
+namespace LaurentSeries
+
+open IsDedekindDomain.HeightOneSpectrum PowerSeries RatFunc
+
+instance : Valued (LaurentSeries K) ℤₘ₀ := Valued.mk' (PowerSeries.idealX K).valuation
+
+theorem valuation_X_pow (s : ℕ) :
+    Valued.v (((X : K⟦X⟧) : LaurentSeries K) ^ s) = Multiplicative.ofAdd (-(s : ℤ)) := by
+  erw [map_pow,/-  this, -/ ← one_mul (s : ℤ), ← neg_mul (1 : ℤ) s, Int.ofAdd_mul,
```
Comment (jcommelin) [inline Mathlib/RingTheory/LaurentSeries.lean:518]:
```suggestion
  erw [map_pow, ← one_mul (s : ℤ), ← neg_mul (1 : ℤ) s, Int.ofAdd_mul,
```

Comment (jcommelin) [reply Mathlib/RingTheory/LaurentSeries.lean:518]: Also, do you understand where exactly you need the `e` in `erw`? Is this pointing at a missing API lemma?

File: Mathlib/RingTheory/LaurentSeries.lean [hunk 3]
```diff
@@ -483,4 +490,168 @@ theorem intValuation_X : (idealX K).intValuationDef X = ↑(Multiplicative.ofAdd
...
+        valuation_X_pow, ofAdd_neg, WithZero.coe_inv, inv_inv]
+  · simp only [Valuation.ne_zero_iff, ne_eq, one_ne_zero, not_false_iff, HahnSeries.single_ne_zero]
+
+/- The coefficients of a power series vanish in degree strictly less than its valuation. -/
+theorem coeff_zero_of_lt_intValuation {n d : ℕ} {f : K⟦X⟧}
+    (H : Valued.v (f : LaurentSeries K) ≤ Multiplicative.ofAdd (-d : ℤ)) :
+    n < d → coeff K n f = 0 := by
+  intro hnd
+  apply (PowerSeries.X_pow_dvd_iff).mp _ n hnd
+  erw [← span_singleton_dvd_span_singleton_iff_dvd, ← Ideal.span_singleton_pow,
```
Comment (jcommelin) [inline Mathlib/RingTheory/LaurentSeries.lean:542]: Same `erw` question.

File: Mathlib/RingTheory/LaurentSeries.lean [hunk 4]
```diff
@@ -483,4 +490,168 @@ theorem intValuation_X : (idealX K).intValuationDef X = ↑(Multiplicative.ofAdd
...
+    ← valuation_of_algebraMap (R := K⟦X⟧) (K := (LaurentSeries K))]
+  exact H
+
+/- The valuation of a power series is the order of the first non-zero coefficient. -/
+theorem intValuation_le_iff_coeff_lt_eq_zero {d : ℕ} (f : K⟦X⟧) :
+    Valued.v (f : LaurentSeries K) ≤ Multiplicative.ofAdd (-d : ℤ) ↔
+      ∀ n : ℕ, n < d → coeff K n f = 0 := by
+  have : PowerSeries.X ^ d ∣ f ↔ ∀ n : ℕ, n < d → (PowerSeries.coeff K n) f = 0 :=
+    ⟨PowerSeries.X_pow_dvd_iff.mp, PowerSeries.X_pow_dvd_iff.mpr⟩
+  erw [← this, valuation_of_algebraMap (PowerSeries.idealX K) f, ←
```
Comment (jcommelin) [inline Mathlib/RingTheory/LaurentSeries.lean:553]: Idem

File: Mathlib/RingTheory/LaurentSeries.lean [hunk 5]
```diff
@@ -483,4 +490,168 @@ theorem intValuation_X : (idealX K).intValuationDef X = ↑(Multiplicative.ofAdd
...
+      rw [← neg_neg (-D + ↑s), ← sub_eq_neg_add, neg_sub, hd]
+      apply (intValuation_le_iff_coeff_lt_eq_zero K F).mpr
+      intro n hn
+      rw [powerSeriesPart_coeff f n, hs]
+      apply h_val_f (s + n)
+      linarith
+    simp only [ne_eq, WithZero.coe_ne_zero, not_false_iff]
+
+/- Two Laurent series whose difference has small valuation have the same coefficients for
+small enough indeces. -/
```
Comment (jcommelin) [inline Mathlib/RingTheory/LaurentSeries.lean:627]:
```suggestion
small enough indices. -/
```

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**